### ISSUE TITLE
Give a better error when using an undeclared macro variable

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#95943]
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a17456ca7139db13e945cbde87ae6fbd

```rust
macro_rules! m {
    () => {
        $x;
    }
}

fn foo() {
    m!()
}
```

The current output is:

```
error: expected expression, found `$`
 --> src/lib.rs:3:9
  |
3 |         $x;
  |         ^^ expected expression
...
8 |     m!()
  |     ---- in this macro invocation
  |
  = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Ideally the output should look like:

```
error[E0425]: cannot find macro parameter `$x` in this scope
 --> src/lib.rs:8:5
  |
2 |    () => {
  |    ^^ in this macro matcher
3 |     $x;
  |     ^^ not found in this scope
...
8 |     m!()
  |     ---- in this macro invocation
  |
  = note: this error originates in the macro `m` (in Nightly builds, run with -Z macro-backtrace for more info)
```

@rustbot label +A-macros

**Comment by PatchMixolydic** [comment]
This scenario is caught by [`meta_variable_misuse`](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#meta-variable-misuse), which almost produces the desired diagnostic except for the span pointing at the macro matchers ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=84fce95a7eaa4b537ecf844e3476ad94)):
```
warning: unknown macro variable `x`
 --> src/lib.rs:5:9
  |
5 |         $x;
  |         ^^
  |
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(meta_variable_misuse)]
  |         ^^^^^^^^^^^^^^^^^^^^
```

It would be beneficial if this lint was documented somewhere a little less obscure than the rustc book, but I'm not sure what a good place would be (maybe the Reference or the unofficial Little Book of Rust Macros?).

cc #61053

**Comment by jyn514** [comment]
It seems strange that's a lint and not just a built-in diagnostic? I'd at least expect it to warn by default ...

**Comment by mark-i-m** [comment]
@jyn514 It's not an error or a warn-by-default because MBE macros are unbelievably permissive, which makes it impossible to detect things like this without false positives. For example, it is hard to determine such a snippet is occurring inside a nested macro definition, in which case the outer macro may be defining the meta-variable.

We spent a lot of time trying to figure out what we can check and how: https://github.com/rust-lang/rust/issues/61053

IMHO, we should remove the MBE system altogether in a future edition and create something more constrained/well-defined that is designed to be (1) easy to document (current macros have terrible documentation), (2) statically checkable, and (3) easy to use. Right now, MBEs satisfy none of these conditions and are not fixable as far as I can tell.

**Comment by jyn514** [comment]
> We spent a lot of time trying to figure out what we can check and how: #61053

I thought the issues you ran into were because you were trying to check it statically at definition time, right? I think giving an error about undefined metavariables should be easier at expansion time.

**Comment by mark-i-m** [comment]
Possibly, but is that useful? As a macro user, what would you do if you found that the macro author made a mistake? I suppose you would just report the bug and hope they fix it. I suppose that's marginally better than getting the error above...

**Comment by jyn514** [comment]
@mark-i-m not all macros are defined cross-crate; IME most of them are local to the current crate because people don't want to expose macros in the API. Having the better error is very useful for local macros, or just when testing your public macro.

**Comment by mark-i-m** [comment]
That's fair enough.

**Comment by mark-i-m** [comment]
It looks like this is the code you would want to change:

https://github.com/rust-lang/rust/blob/3bebee73397b692e84b9cc1b6af439fae78918c9/compiler/rustc_expand/src/mbe/transcribe.rs#L249-L254

I'm not sure how you could tell if you were in a nested invocation or not, though. One option is to just make it an error and do a crater run to see how wide the impact is.

Maybe the AST Validation might be another place to check for this? I don't fully understand what happens in that pass, but it happens after all macros are fully expanded.

**Comment by anoshyn** [comment]
I'd like to take this on. @mark-i-m's 2022 pointer to `compiler/rustc_expand/src/mbe/transcribe.rs:249-254` is the right area but the lines have shifted (the file is now 4× larger; the relevant function is `transcribe_metavar` at line 424). Since the original mentor is no longer active, would anyone on T-compiler be available to provide guidance? Happy to also ask in `#t-compiler/help` on Zulip.

@rustbot claim

**Comment by anoshyn** [comment]
Unclaiming — I jumped in without sufficient understanding of the macro-expansion area; someone with experience there should take this on.

@rustbot release-assignment
